### PR TITLE
feat: add tests to check for changed files

### PR DESCRIPTION
**Pull request by pimlie**
As discussed in #3869

Make sure that if we are building or generating only files in buildDir and generate.dir are changed. If files in another location would also be changed due to a new config option, those locations should be guarded in `lib/common/options.js` so you cant set them lower then rootDir or srcDir.

## Checklist:
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by pimlie** [comment]
Yay, all checks passed. I did a forced push with the double 'add debug logs' commits to try to remove / hide some commits as I was remotely debugging the CI failures. Either my pc is much slower or much faster then the CI's as locally the writeFile just finished long before running the tests ;)

File: test/utils/build.js [hunk 1]
```diff
@@ -15,4 +24,17 @@ export const buildFixture = function (fixture, callback, hooks = []) {
...
   }, 120000)
+
+  test('Check changed files', () => {
+    expect.hasAssertions()
+
+    // When building Nuxt we only expect files to changed
+    // within the nuxt.options.buildDir
+    Object.keys(pathsBefore).forEach((key) => {
+      const paths = listPaths(nuxt.options[`${key}Dir`], pathsBefore[key])
+      paths.map((item) => {
```
Comment (manniL) [inline test/utils/build.js]: `forEach` instead of `map` if you don't need a return value.

File: test/unit/basic.generate.test.js [hunk 2]
```diff
@@ -43,6 +53,24 @@ describe('basic generate', () => {
...
 
+  test('Check changed files', () => {
+    // When generating Nuxt we only expect files to change
+    // within nuxt.options.generate.dir, but also allow other
+    // .nuxt dirs for when tests are runInBand
+    const allowChangesDir = resolve(generator.nuxt.options.generate.dir, '..', '.nuxt')
+
+    let changedFileFound = false
+    const paths = listPaths(generator.nuxt.options.rootDir, pathsBefore)
+    paths.map((item) => {
```
Comment (manniL) [inline test/unit/basic.generate.test.js]: `forEach` instead of `map` if you don't need a return value.

**Review by TheAlexLichter — CHANGES_REQUESTED** [review]

**Comment by pimlie** [comment]
It seems appveyor really screws up the timeout time. I was confident that using a high resolution time would be a quick fix for the waitfor test but as many times before I was wrong. Rounding the time to a whole number fixes the test in this case, but we will have to wait and see if timeouts on appveyor are off by more than 0.5ms. If so we can skip the test again, or maybe file an report with appveyor? Spent way too much on this already so actually I wouldnt mind to understand why this is at all possible ;)

File: test/unit/utils.test.js [hunk 3]
```diff
@@ -16,10 +16,15 @@ describe('utils', () => {
...
-    expect(Date.now() - s >= 100).toBe(true)
+  test('waitFor', async () => {
+    const delay = 100
+    const s = process.hrtime()
+    await Utils.waitFor(delay)
+    const t = process.hrtime(s)
+    // Node.js makes no guarantees about the exact timing of when callbacks will fire
+    // HTML5 specifies a minimum delay of 4ms for timeouts
+    // although arbitrary, use this value to determine our lower limit
+    expect((t[0] * 1e9 + t[1]) / 1e6).not.toBeLessThan(delay - 4)
```
Comment (pi0) [inline test/unit/utils.test.js]: Crazy boy 😂😋

**Review by pi0 — APPROVED** [review]
LGTM

**Review by TheAlexLichter — APPROVED** [review]

File: test/unit/utils.test.js (hunk 3 above)
Comment (clarkdo) [reply test/unit/utils.test.js]: Good idea

**Review by clarkdo — APPROVED** [review]
LGTM

**Pull request merged**